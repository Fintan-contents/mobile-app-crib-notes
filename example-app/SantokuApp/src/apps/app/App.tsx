import {setHandleError} from 'bases/core/errors/handleError';
import {setHandleErrorWithAlert} from 'bases/core/errors/handleErrorWithAlert';
import {setHandleErrorWithSnackbar} from 'bases/core/errors/handleErrorWithSnackbar';
import {firebaseConfig} from 'bases/firebase/FirebaseConfig';
import {createLogger, setLogger} from 'bases/logging';
import {FirebaseCrashlyticsTransport} from 'bases/logging/FirebaseCrashlyticsTransport';
import {LoadingOverlay} from 'bases/ui/loading/LoadingOverlay';
import {Snackbar} from 'bases/ui/snackbar/Snackbar';
import {AppThemeProvider} from 'bases/ui/theme/AppThemeProvider';
import React from 'react';
import {Platform, StyleSheet} from 'react-native';
import {GestureHandlerRootView} from 'react-native-gesture-handler';
import {SafeAreaProvider} from 'react-native-safe-area-context';

import {AppWithInitialization} from './AppWithInitialization';
import {handleError} from './errors/handleError';
import {handleErrorWithAlert} from './errors/handleErrorWithAlert';
import {handleErrorWithSnackbar} from './errors/handleErrorWithSnackbar';

type AppProperties = {
  [key: string]: any;
};

setLogger(
  __DEV__ || firebaseConfig.isDummy
    ? createLogger({level: 'trace'})
    : createLogger({level: 'error', transports: [new FirebaseCrashlyticsTransport()]}),
);

setHandleError(handleError);
setHandleErrorWithSnackbar(handleErrorWithSnackbar);
setHandleErrorWithAlert(handleErrorWithAlert);

export const App = ({isHeadless}: AppProperties) => {
  // isHeadlessを取得するためにはAppDelegate.mに変更が必要
  // https://rnfirebase.io/messaging/usage#background-application-state
  if (Platform.OS === 'ios' && isHeadless) {
    // iOSでバックグラウンド時にリモート通知を受信するとアプリケーションがサイレントに起動される
    // この時にはReact Componentを返さないようにし、useEffectなどの処理がバックグラウンドで走ることを防止する
    // Androidの場合はApp自体が読み込まれないので追加対応は必要ない
    return null;
  }

  return (
    <GestureHandlerRootView style={StyleSheet.absoluteFill}>
      <SafeAreaProvider>
        <AppThemeProvider>
          <AppWithInitialization />
          <LoadingOverlay.Component />
          <Snackbar.Component />
        </AppThemeProvider>
      </SafeAreaProvider>
    </GestureHandlerRootView>
  );
};
