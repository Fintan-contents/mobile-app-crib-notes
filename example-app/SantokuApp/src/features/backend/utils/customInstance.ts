import Axios, {AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse} from 'axios';
import {AppConfig} from 'bases/core/configs/AppConfig';
import {applicationName, nativeApplicationVersion} from 'expo-application';
import {RequestTimeoutError} from 'features/backend/errors/RequestTimeoutError';
import {Platform} from 'react-native';

export type ErrorType<Error> = AxiosError<Error>;

const REQUEST_TIMEOUT = AppConfig.requestTimeout;

const backendUrl = `${AppConfig.santokuAppBackendUrl}/api`;
const sandboxUrl = `${AppConfig.santokuAppBackendUrl}/api/sandbox`;
const BACKEND_AXIOS_INSTANCE = Axios.create({baseURL: backendUrl});
const SANDBOX_AXIOS_INSTANCE = Axios.create({baseURL: sandboxUrl});
const BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION = Axios.create({baseURL: backendUrl});

const getUserAgent = () => {
  const appName = applicationName ?? 'unknown';
  const appVersion = nativeApplicationVersion ?? 'unknown';
  const platform = Platform.OS;
  const platformVersion = Platform.Version;
  return `${appName}/${appVersion} (${platform}/${platformVersion})`;
};

const getDefaultAxiosConfig = () => {
  return {
    headers: {
      Accept: 'application/json',
      UserAgent: getUserAgent(),
    },
  } as AxiosRequestConfig;
};

const customInstance = <T>(
  axiosInstance: AxiosInstance,
): ((config: AxiosRequestConfig) => Promise<AxiosResponse<T>>) => {
  const defaultAxiosConfig = getDefaultAxiosConfig();
  return (config: AxiosRequestConfig) => {
    // TODO: React Native / Expo のバージョンアップ時にJestを27以降にバージョンアップできたらCancelTokenからAbortControllerへ移行する
    const source = Axios.CancelToken.source();
    const requestConfig = {
      ...defaultAxiosConfig,
      ...config,
      cancelToken: source.token,
    };
    const promise = axiosInstance(requestConfig);

    // @ts-ignore
    promise.cancel = () => {
      source.cancel('Query was cancelled by React Query');
    };

    let timeoutId: NodeJS.Timeout | null;
    if (REQUEST_TIMEOUT) {
      timeoutId = setTimeout(() => {
        timeoutId = null;
        source.cancel('Query was cancelled by Request Timeout');
      }, REQUEST_TIMEOUT);
    }

    return promise
      .catch(error => {
        // TODO: AbortControllerへの移行時に処理を見直す
        // AbortControllerでabortした場合はErrorにはならず、responseがundefinedになる
        if (Axios.isCancel(error)) {
          const cancelError = error as {message: string};
          if (cancelError.message === 'Query was cancelled by Request Timeout') {
            throw new RequestTimeoutError('Request Timeout');
          }
        }
        throw error;
      })
      .finally(() => {
        if (timeoutId) {
          clearTimeout(timeoutId);
        }
      });
  };
};

const backendCustomInstance = <T>(config: AxiosRequestConfig): Promise<AxiosResponse<T>> => {
  return customInstance<T>(BACKEND_AXIOS_INSTANCE)(config);
};

const sandboxCustomInstance = <T>(config: AxiosRequestConfig): Promise<AxiosResponse<T>> => {
  return customInstance<T>(SANDBOX_AXIOS_INSTANCE)(config);
};

const setCsrfTokenHeader = (csrfTokenHeaderName: string, csrfTokenValue: string) => {
  BACKEND_AXIOS_INSTANCE.defaults.headers.common[csrfTokenHeaderName] = csrfTokenValue;
  BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION.defaults.headers.common[csrfTokenHeaderName] = csrfTokenValue;
};

const setAxiosResponseInterceptor = (
  onFulfilled: (
    value: AxiosResponse<any, any>,
  ) => (AxiosResponse<any, any> | Promise<AxiosResponse<any, any>>) | undefined,
  onRejected: (error: any) => any | undefined,
) => {
  BACKEND_AXIOS_INSTANCE.interceptors.response.use(onFulfilled, onRejected);
};

export {
  backendCustomInstance,
  sandboxCustomInstance,
  setCsrfTokenHeader,
  setAxiosResponseInterceptor,
  BACKEND_AXIOS_INSTANCE,
  BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION,
};
