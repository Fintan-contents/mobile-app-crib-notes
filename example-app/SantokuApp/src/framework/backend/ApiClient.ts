import axios, {AxiosError} from 'axios';

import {AccountApi, ErrorResponse, SystemApi, TeamApi, TermsApi} from '../../generated/api';
import {AppConfig} from '../config';
import {ApplicationError} from '../error/ApplicationError';

/** APIでエラーが発生した場合に送出するエラーです。 */
export class ApiError extends ApplicationError {}

/** APIでエラー発生時に、レスポンスが存在する場合のエラーです。 */
export class ApiResponseError extends ApiError {
  constructor(
    cause: AxiosError,
    public response: {
      data: ErrorResponse;
      status: number;
      statusText: string;
      headers: Record<string, string>;
      request?: any;
    },
  ) {
    super(cause);
  }
}

/** APIでエラー発生時に、レスポンスが存在しない場合のエラーです。 */
export class ApiResourceAccessError extends ApiError {}

const instance = axios.create();
instance.interceptors.response.use(
  (response) => response,
  (error) => {
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError<ErrorResponse>;
      if (axiosError.response) {
        throw new ApiResponseError(axiosError, axiosError.response);
      }
      throw new ApiResourceAccessError(error);
    }
    throw new ApiError(error);
  },
);

export const systemApi = new SystemApi(undefined, `${AppConfig.santokuAppBackendUrl}/api`, instance);
export const accountApi = new AccountApi(undefined, `${AppConfig.santokuAppBackendUrl}/api`, instance);
export const termsApi = new TermsApi(undefined, `${AppConfig.santokuAppBackendUrl}/api`, instance);
export const teamApi = new TeamApi(undefined, `${AppConfig.santokuAppBackendUrl}/api`, instance);

export async function csrfToken() {
  const token = await systemApi.getCsrfToken();
  const csrfTokenValue = token.data.csrfTokenValue;
  const csrfTokenHeaderName = token.data.csrfTokenHeaderName;
  instance.defaults.headers.common[csrfTokenHeaderName] = csrfTokenValue;
}
