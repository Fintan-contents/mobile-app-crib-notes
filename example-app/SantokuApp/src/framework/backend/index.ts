export * from './refreshCsrfToken';
export * from './refreshSession';
