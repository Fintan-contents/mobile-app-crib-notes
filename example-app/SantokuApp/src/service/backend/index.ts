export * from './accountService';
export * from './systemService';
export * from './termService';
export * from './sandboxService';
