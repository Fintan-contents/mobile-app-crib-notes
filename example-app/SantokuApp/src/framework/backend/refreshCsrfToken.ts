import {BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION, setCsrfTokenHeader} from './customInstance';

type CsrfTokenResponse = {
  csrfTokenHeaderName: string;
  csrfTokenValue: string;
};

const refreshCsrfToken = async () => {
  const axiosResponse = await BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION.get<CsrfTokenResponse>('/csrf_token');
  setCsrfTokenHeader(axiosResponse.data.csrfTokenHeaderName, axiosResponse.data.csrfTokenValue);
};

export {refreshCsrfToken};
