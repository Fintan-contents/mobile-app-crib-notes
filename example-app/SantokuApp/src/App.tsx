import {NavigationContainer} from '@react-navigation/native';
import {activateKeepAwake} from 'expo-keep-awake';
import {RootStackNav} from 'navigation';
import React from 'react';

import {BundledMessagesLoader, loadMessages} from './framework';
import {firebaseConfig} from './framework/firebase';
import {log} from './framework/logging';
import {launchedId} from './framework/utilities/id';
import crashlytics from '@react-native-firebase/crashlytics';

export const App = () => {
  // 開発中は画面がスリープしないようにしておきます。
  if (__DEV__) {
    activateKeepAwake();
  }

  if (!firebaseConfig.isDummy) {
    // Firebase Crashlyticsの初期化
    // アプリの起動IDを設定
    launchedId()
      .then((value) => crashlytics().setAttribute('launchedId', value))
      .catch(() => {
        log.error('Failed to get launchedId.', 'LaunchedIdGettingFailure');
      });
  }
  // アプリ内で使用するメッセージのロード
  loadMessages(new BundledMessagesLoader()).catch(() => {
    // アプリにバンドルしているメッセージのロードは失敗しない想定
    log.error('Failed to load message.', 'BundledMessagesLoadingFailure');
  });

  return (
    <NavigationContainer>
      <RootStackNav />
    </NavigationContainer>
  );
};
