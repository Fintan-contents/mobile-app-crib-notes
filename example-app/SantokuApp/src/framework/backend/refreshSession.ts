import axios, {AxiosResponse} from 'axios';

import {SecureStorageAdapter} from '../authentication';
import {ApplicationError} from '../error/ApplicationError';
import {sendErrorLog} from '../error/sendErrorLog';
import {BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION, setAxiosResponseInterceptor} from './customInstance';
import {refreshCsrfToken} from './refreshCsrfToken';

// 共通リトライ処理に必要なAPI呼び出しだけは、自動生成コードに依存しない形で用意する
const autoLogin = async () => {
  const accountId = await SecureStorageAdapter.loadActiveAccountId();
  if (!accountId) {
    throw new ApplicationError('There is no auto-login enabled account.');
  }
  const password = await SecureStorageAdapter.loadPassword(accountId);
  if (!password) {
    throw new ApplicationError('The password for the account ID does not exist.');
  }
  await BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION.post('/login', {accountId, password});
};

const refreshSession = async () => {
  await autoLogin();
  await refreshCsrfToken();
};

const setRefreshSessionInterceptor = () => {
  const onFulfilled = (response: AxiosResponse) => {
    return response;
  };
  const onRejected = async (error: unknown) => {
    if (axios.isAxiosError(error)) {
      if (error.response?.status === 401) {
        try {
          await refreshSession();
          return await BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION.request(error.config);
        } catch (e) {
          sendErrorLog(e);
          throw error;
        }
      }
    }
    throw error;
  };
  setAxiosResponseInterceptor(onFulfilled, onRejected);
};

export {refreshSession, setRefreshSessionInterceptor};
