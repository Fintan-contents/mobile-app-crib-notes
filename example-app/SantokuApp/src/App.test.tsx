import '@testing-library/jest-native/extend-expect';
import {render, waitFor} from '@testing-library/react-native';
import React from 'react';
import {DevSettings} from 'react-native';

import {App} from './App';
import {BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION} from './framework/backend/customInstance';

jest.spyOn(DevSettings, 'addMenuItem').mockImplementation(() => {});

beforeEach(() => {
  // 画面遷移時のアニメーションが、コンポーネントのアンマウント後にステートを更新してしまうようで、
  // テストは成功するものの、エラーログが出力されてしまう。
  // タイマーを使わないようにして、アニメーションを動かさないことで回避しているつもり。
  jest.useFakeTimers();
});

describe('App', () => {
  it('マウントされたときに正常にレンダリングされること', async () => {
    jest.spyOn(BACKEND_AXIOS_INSTANCE_WITHOUT_REFRESH_SESSION, 'get').mockResolvedValue({
      status: 200,
      data: {
        csrfTokenHeaderName: 'X-CSRF-TOKEN',
        csrfTokenValue: 'dummy',
      },
    });
    const app = render(<App />);
    await waitFor(() => {
      expect(app.queryByTestId('TermsOfServiceAgreementScreen')).not.toBeNull();
      expect(app).toMatchSnapshot();
    });
  });
});
