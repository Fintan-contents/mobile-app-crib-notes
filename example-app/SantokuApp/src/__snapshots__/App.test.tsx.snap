// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App マウントされたときに正常にレンダリングされること 1`] = `
Array [
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        enabled={true}
        isNativeStack={true}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        stackPresentation="push"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          backButtonInCustomView={false}
          backgroundColor="rgb(255, 255, 255)"
          color="rgb(0, 122, 255)"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={false}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="利用規約"
          titleColor="rgb(28, 28, 30)"
          topInsetEnabled={false}
          translucent={false}
        />
        <View
          style={
            Array [
              Object {
                "flex": 1,
              },
              Object {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "white",
                "flex": 1,
                "justifyContent": "center",
              }
            }
            testID="TermsOfServiceAgreementScreen"
          >
            <View
              style={
                Array [
                  Object {
                    "flex": 1,
                    "overflow": "hidden",
                  },
                  undefined,
                ]
              }
            >
              <RNCWebView
                androidLayerType="software"
                cacheEnabled={true}
                decelerationRate={0.998}
                injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
                injectedJavaScriptForMainFrameOnly={true}
                javaScriptEnabled={true}
                messagingEnabled={false}
                onContentProcessDidTerminate={[Function]}
                onError={[Function]}
                onHttpError={[Function]}
                onLoadProgress={[Function]}
                onLoadStart={[Function]}
                onLoadingError={[Function]}
                onLoadingFinish={[Function]}
                onLoadingProgress={[Function]}
                onLoadingStart={[Function]}
                onMessage={[Function]}
                onScroll={[Function]}
                onShouldStartLoadWithRequest={[Function]}
                source={
                  Object {
                    "uri": "https://www.tis.co.jp/termsofuse/",
                  }
                }
                startInLoadingState={true}
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Object {
                      "backgroundColor": "#ffffff",
                    },
                    Object {
                      "flex": 1,
                    },
                  ]
                }
                useSharedProcessPool={true}
              />
              <ActivityIndicator
                animating={true}
                color="#0000ff"
                hidesWhenStopped={true}
                size="large"
                style={
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "height": "100%",
                    "justifyContent": "center",
                    "position": "absolute",
                    "width": "100%",
                  }
                }
              />
            </View>
            <View
              style={
                Object {
                  "flex": 0,
                  "flexDirection": "row-reverse",
                  "paddingHorizontal": 20,
                  "paddingVertical": 14,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "borderRadius": 3,
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 3,
                    },
                    Object {
                      "borderRadius": 8,
                      "width": 90,
                    },
                    false,
                  ]
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    Object {
                      "busy": false,
                      "disabled": true,
                    }
                  }
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "opacity": 1,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "hsl(208, 8%, 90%)",
                        "borderColor": "#2089dc",
                        "borderRadius": 8,
                        "borderWidth": 0,
                        "flexDirection": "row",
                        "height": 50,
                        "justifyContent": "center",
                        "padding": 8,
                        "width": 90,
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "color": "hsl(208, 8%, 63%)",
                          "fontSize": 18,
                          "paddingVertical": 1,
                          "textAlign": "center",
                        }
                      }
                    >
                      同意
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaProvider>,
  <View
    testID="FullWindowOverlay"
  />,
]
`;
